lib/aws/request.js

```diff
@@ -0,0 +1,218 @@
+  sdk.config.httpOptions.agent = new https.Agent(proxyOptions);
+}
+
+// Configure the AWS Client timeout (Optional).  The default is 120000 (2 minutes)
+const timeout = process.env.AWS_CLIENT_TIMEOUT || process.env.aws_client_timeout;
+if (timeout) {
+  sdk.config.httpOptions.timeout = parseInt(timeout, 10);
+}
+PromiseQueue.configure(Promise);
+const requestQueue = new PromiseQueue(2, Infinity);
+
+const MAX_RETRIES = (() => {
+  const userValue = Number(process.env.SLS_AWS_REQUEST_MAX_RETRIES);
+  return userValue >= 0 ? userValue : 4;
+})();
+
+const accelerationCompatibleS3Methods = new Set(['upload', 'putObject']);
+
+const getServiceInstance = memoize(
+  (service, method) => {
+    const Service = _.get(sdk, service.name);
+    // Support S3 Transfer Acceleration helper function
+    const shouldS3Accelerate = () => {
+      if (
+        accelerationCompatibleS3Methods.has(method) &&
+        service.params.isS3TransferAccelerationEnabled
+      ) {
+        log('Using S3 Transfer Acceleration Endpoint...');
+        return true;
+      }
+      return false;
+    };
+    // we translate params to an object for the service creation by selecting keys of interest
+    const serviceParams = {
+      credentials: service.params.credentials,
+      region: service.params.region ? service.params.region : 'us-east-1',
+    };
+    // we need to take s3 acceleration options into account
+    if (service.name === 'S3') serviceParams.useAccelerateEndpoint = shouldS3Accelerate();
+    return new Service(serviceParams);
+  },
+  {
+    normalizer: ([service, method]) => {
+      return [JSON.stringify(deepSortObjectByKey(service)), method].join('|');
+    },
+  }
+);
+
+/** Execute request to AWS service
+ * @param {Object} [service] - Description of the service to call
+ * @prop [service.name] - Name of the service to call, support subclasses
+ * @prop [service.params] - Parameters to apply when creating the service and doing the request
+ * @prop [service.params.credentials] - AWS Credentials to use
+ * @prop [service.params.useCache ] - Wether to reuse result of the same request cached locally
+ * @prop [service.params.region] - Region in which the call should be made (default to us-east-1)
+ * @prop [service.params.isS3TransferAccelerationEnabled] - Use s3 acceleration when available for the request
+ * @param {String} method - Method to call
+ * @param {Array} args - Argument for the method call
+ */
+function awsRequest(service, method, ...args) {
+  // Checks regarding expectations on service object
+  ensureObject(service, { name: 'service', Error: ServerlessError });
+  ensureObject(service.params, { name: 'service.params', Error: ServerlessError });
+  ensureObject(service.params.credentials, {
+    name: 'service.params.credentials',
+    Error: ServerlessError,
+  });
+  ensureString(service.name, { name: 'service.name', Error: ServerlessError });
```

Comment:
Does that mean that we should support making request without authentication?

Reply:
If not provided with params, AWS SDK reads authentication on its own internally reading from env vars.

I think it's nice to leave an option for this util to be used that way. (by secluding, it's no longer tight to `Serverless` instance, so it could be used e.g. by scripts)

Also it's possible that in context of a Framework we will refactor internal credentials handling, so it'll be not passed with params, but set directly on `process.env` (to be discussed)

Reply:
Ok, this is a bigger discussion indeed :smile:  I'll remove the constraint there and try to support the service/string use case with associated test.

Reply:
Unfortunately this lead me to remove most of the authentication checks. The failure were happening when creating the service and not later on. I'd like to write the authentication scenario and failure mode for that awsRequest function if we want to introduce a bit more checking.

Reply:
> Unfortunately this lead me to remove most of the authentication checks.

Do you mean checks that confirm that those options are validated. Or some other  checks?

Reply:
Other checks, including the one redirecting to serverless documentation for authentication issues.

Reply:
I've just posted review, I believe we really should keep that part. What was the reason you've decided to strip it?

I think all it does, is making an authentication error communicated with way better message, than if we surface one from AWS directly